### PR TITLE
Fake auditor runtime option

### PR DESCRIPTION
**Pull request by blue42u**
Moves the configure-time option for LD_AUDIT to hpcrun-time. To be added to the "use when it crashes" option set.

### CODE REVIEW

**Comment by mxz297** [comment]
@jmellorcrummey @mwkrentel  @blue42u I have tested this PR with a new unit test added to the hpctoolkit-test repository (https://github.com/HPCToolkit/hpctoolkit-tests/tree/master/unit-test/runpath). This unit test finds a problem in the hpcrun script that accidentally adds the current directory to LD_LIBRARY_PATH; this problem also exists in the develop branch. @blue42u has pushed a new commit that fixs this problem in this PR.

Through my testing, I find that when spack builds libmonitor, libmonitor will still contain dlopen overrides. This basically defeats the purpose of using rtld-audit. @blue42u told me that I needed to build libmonitor outside spack with `--disable-dlfcn`. With the libmonitor that does not have dlopen, the unit test works as expected.

I think this PR is good for merging but we need to also fix the libmonitor through spack. With the fake auditor providing function overrides, can we make `--disable-dlfcn` the default option for building libmonitor?

**Comment by mwkrentel** [comment]
> can we make --disable-dlfcn the default option for building
> libmonitor?

No, we don't need to make that the default.  There already is a knob
to set it one way or the other.

The right solution is to push the knob to the spack level and then
hpctoolkit should set the option to on/off depending on the rest of
the build.

I'll add the knob to monitor sometime during vacation time. :-)
I just want to finish up some rocm things first.

File: src/tool/hpcrun/scripts/hpcrun.in [hunk 1]
```diff
@@ -503,10 +508,21 @@ fi
...
+hpc_ld_library_path="${hpc_ld_library_path}${perfmon_libdir:+:${perfmon_libdir}}"
+hpc_ld_library_path="${hpc_ld_library_path}${roctracer_lib:+:${roctracer_lib}}"
 preload_list="${libmonitor_dir}/libmonitor.so ${preload_list}"
 preload_list="${hpcrun_dir}/libhpcrun.so ${preload_list}"
 
+# Set up the auditor or fake auditor, depending on what's configured.
+if test "$HPCRUN_AUDIT_USE_FAKE_AUDITOR" = 1 ; then
+    preload_list="${hpcrun_dir}/libhpcrun_fake_audit.so ${preload_list}"
+else
+    audit_list="${hpcrun_dir}/libhpcrun_audit.so:${audit_list}"
```
Comment (jmellorcrummey) [inline src/tool/hpcrun/scripts/hpcrun.in:523]: Does this line need to be

audit_list="${hpcrun_dir}/libhpcrun_audit.so${audit_list:+:${audit_list}}"

File: src/tool/hpcrun/scripts/hpcrun.in [hunk 2]
```diff
@@ -520,9 +536,9 @@ if test ! -x "$HPCRUN_FNBOUNDS_CMD" ; then
 fi
 export HPCRUN_FNBOUNDS_CMD
 
-export LD_LIBRARY_PATH="${hpc_ld_library_path}:${LD_LIBRARY_PATH}"
+export LD_LIBRARY_PATH="${hpc_ld_library_path}${LD_LIBRARY_PATH:+:${LD_LIBRARY_PATH}}"
 export LD_PRELOAD="${preload_list} ${LD_PRELOAD}"
-@export_ld_audit@
+export LD_AUDIT="${audit_list}:${LD_AUDIT}"
```
Comment (jmellorcrummey) [inline src/tool/hpcrun/scripts/hpcrun.in:541]: What if either of these terms is NULL? Is the leading or trailing : something that should be avoided? Do we even want to set LD_AUDIT if both LD_AUDIT and audit_list are empty?

File: src/tool/hpcrun/scripts/hpcrun.in [hunk 3]
```diff
@@ -503,10 +508,21 @@ fi
...
+hpc_ld_library_path="${hpcrun_dir}"
+hpc_ld_library_path="${hpc_ld_library_path}${papi_libdir:+:${papi_libdir}}"
+hpc_ld_library_path="${hpc_ld_library_path}${ext_dir:+:${ext_dir}}"
+hpc_ld_library_path="${hpc_ld_library_path}${perfmon_libdir:+:${perfmon_libdir}}"
+hpc_ld_library_path="${hpc_ld_library_path}${roctracer_lib:+:${roctracer_lib}}"
 preload_list="${libmonitor_dir}/libmonitor.so ${preload_list}"
 preload_list="${hpcrun_dir}/libhpcrun.so ${preload_list}"
 
+# Set up the auditor or fake auditor, depending on what's configured.
+if test "$HPCRUN_AUDIT_USE_FAKE_AUDITOR" = 1 ; then
```
Comment (jmellorcrummey) [inline src/tool/hpcrun/scripts/hpcrun.in:520]: I prefer the shorter HPCRUN_FAKE_AUDITOR (here and in the code)

**Review by jmellorcrummey — APPROVED** [review]
My comments suggest a few changes. Other than those, everything looks fine.

File: src/tool/hpcrun/scripts/hpcrun.in (hunk 1 above)
Comment (blue42u) [reply src/tool/hpcrun/scripts/hpcrun.in]: Empty entries in the LD_AUDIT and LD_PRELOAD line don't do anything, but changed for consistency and cleanliness.

File: src/tool/hpcrun/scripts/hpcrun.in (hunk 2 above)
Comment (blue42u) [reply src/tool/hpcrun/scripts/hpcrun.in]: Acts as a no-op, but fixed anyway.

File: src/tool/hpcrun/scripts/hpcrun.in (hunk 3 above)
Comment (blue42u) [reply src/tool/hpcrun/scripts/hpcrun.in]: Done.

**Comment by mwkrentel** [comment]
If an empty path component really is a problem for LD_LIBRARY_PATH,
then I'd suggest just normalizing the value.  Otherwise, you're at the
mercy of the value from the user and you'll end up answering more
questions and blaming users.  Easier to make it "just work."

Here is a straightforward way to do this (with my test program).

```
#!/bin/sh

STR='::abc::/path/to/dir::.:dir  with  spaces::end:::'
echo "$STR"

OLDIFS="$IFS"
IFS=':;'
new_path=
for elt in $STR ; do
    if test "x$elt" != x ; then
	case "$new_path" in
	    '' ) new_path="$elt" ;;
	    * )  new_path="${new_path}:$elt" ;;
	esac
    fi
done
IFS="$OLDIFS"

echo "$new_path"
```

Man ld.so says that LD_LIBRARY_PATH uses either colon or semi-colon
for delimiters, so I guess to use `IFS=':;'`

Btw, empty elements are not supposed to happen in the for loop.  I
found that they don't happen with `IFS=' '` but do with `IFS=':'` which is
in direct contradiction to Blinn.  I'd guess that bash is incorrect here.

**Comment by jmellorcrummey** [comment]
> If an empty path component really is a problem for LD_LIBRARY_PATH,
> then I'd suggest just normalizing the value. Otherwise, you're at the
> mercy of the value from the user and you'll end up answering more
> questions and blaming users. Easier to make it "just work."
> 
> Here is a straightforward way to do this (with my test program).
> 
> ```
> #!/bin/sh
> 
> STR='::abc::/path/to/dir::.:dir  with  spaces::end:::'
> echo "$STR"
> 
> OLDIFS="$IFS"
> IFS=':;'
> new_path=
> for elt in $STR ; do
>     if test "x$elt" != x ; then
> 	case "$new_path" in
> 	    '' ) new_path="$elt" ;;
> 	    * )  new_path="${new_path}:$elt" ;;
> 	esac
>     fi
> done
> IFS="$OLDIFS"
> 
> echo "$new_path"
> ```
> 
> Man ld.so says that LD_LIBRARY_PATH uses either colon or semi-colon
> for delimiters, so I guess to use `IFS=':;'`
> 
> Btw, empty elements are not supposed to happen in the for loop. I
> found that they don't happen with `IFS=' '` but do with `IFS=':'` which is
> in direct contradiction to Blinn. I'd guess that bash is incorrect here.

I don't think we should remove empty entries from a user's LD_LIBRARY_PATH by sanitization. Their path should be whatever it was. The only thing we MUST do is to avoid introducing empty path entries (::) that can change the semantics of their (and our) library search results.

**Comment by mwkrentel** [comment]
So, if a user's LD_LIBRARY_PATH contains '::', then you're content to
have the auditor fail?

What's the harm?  I've never heard of this treatment before and it
came as quite a surprise.  I expect there are plenty of users who have
'::' in various path-like variables and have no idea of the side
effects.

**Comment by jmellorcrummey** [comment]
> So, if a user's LD_LIBRARY_PATH contains '::', then you're content to
> have the auditor fail?
> 
> What's the harm? I've never heard of this treatment before and it
> came as quite a surprise. I expect there are plenty of users who have
> '::' in various path-like variables and have no idea of the side
> effects.

You misunderstand. The auditor will not fail. Xiaozhu noted a failure in a regression test that depended upon hpcrun not introducing the current directory as a new LD_LIBRARY_PATH entry. having hpcrun do that caused his regression test to behave differently. That was the only problem we are trying to solve by not spuriously adding :: to the path in hpcrun. If it is already in a user's LD_LIBRARY_PATH, then their application will behave as they expect when measured with hpcrun. Nothing will be different. If we rip something out of a user's LD_LIBRARY_PATH that they introduced intentionally and that breaks their application, that is a bad behavior for hpcrun as well.

**Pull request merged**